### PR TITLE
[DeepSpeed] Do not fail if batch size could not be inferred for logging

### LINKED ISSUE #10437: DeepSpeed should not fail if batch size cannot be inferred
## 🐛 Bug

Currently if DeepSpeed is not able to infer the batch size from the dataloader, this fails. To get around this, you need to instantiate the plugin and pass the batch size via `strategy=DeepSpeedPlugin(logging_batch_size_per_gpu=1)`.

The annoying thing is, this variable is only used in DeepSpeed for internal batch size logging. We've by default turned this off, so users do not even see these logging messages unless they are enabled.

As a result I propose we set this to 1 if we cannot infer automatically and the user has not passed it in. We should throw a warning in case.

### To Reproduce

In the below example, we define our dataloader in `setup` (the dataset is only setup in the `setup` function, in this case I just define the entire loader). This fails till you change `DeepSpeedPlugin()` to `DeepSpeedPlugin(logging_batch_size_per_gpu=1)`.
```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer, LightningDataModule
from pytorch_lightning.plugins import DeepSpeedPlugin


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


class DataModule(LightningDataModule):
    def setup(self, stage=None) -> None:
        self._dataloader = DataLoader(RandomDataset(32, 64), batch_size=2)

    def train_dataloader(self):
        return self._dataloader

    def val_dataloader(self):
        return self._dataloader


def run():
    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        gpus=1,
        limit_train_batches=1,
        limit_val_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
        strategy=DeepSpeedPlugin()
    )
    trainer.fit(model, datamodule=DataModule())


if __name__ == "__main__":
    run()
```
### Expected behavior

Sets the value to 1 and runs.

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Closes #10437 

Adds a try/catch and a check to ensure we only warn if needs be (if logging has been set lower than WARNING).

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 1]
```diff
@@ -634,9 +627,17 @@ def _auto_select_batch_size(self):
         batch_size = 1
         train_dl_source = self.lightning_module.trainer._data_connector._train_dataloader_source
         if train_dl_source.is_defined():
-            train_dataloader = train_dl_source.dataloader()
-            if hasattr(train_dataloader, "batch_sampler"):
-                batch_size = train_dataloader.batch_sampler.batch_size
+            try:
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/deepspeed.py:630]: Only wrap what you expect to fail, I assuming it's only L633

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 2]
```diff
@@ -634,9 +627,17 @@ def _auto_select_batch_size(self):
...
+                train_dataloader = train_dl_source.dataloader()
+                if hasattr(train_dataloader, "batch_sampler"):
+                    batch_size = train_dataloader.batch_sampler.batch_size
+            except Exception:
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/deepspeed.py:634]: can't it be entirely resolved with two chained hasattr checks?

[the remark this answers is not in the document]
Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:634]: @carmocca any suggestions on what to narrow it to? I'm uncertain what types it can be.

@awaelchli I'm not sure, `train_dl_source.dataloader()` from my understanding just calls the reference datamodule or hook in the lightning object. This can fail in unexpected ways if `setup` has not been called (Which is not a hard requirement, but is for certain DataModules).

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 3]
```diff
@@ -462,9 +462,7 @@ def init_deepspeed(self):
         if self.zero_stage_3 and self.partition_module:
             # Ensure the entire model has been moved to the appropriate device
             dtype = torch.float16 if self.precision in (16, "mixed") else torch.float32
-            deepspeed.zero.Init(
-                module=model, remote_device=self.remote_device, pin_memory=True, config=self.config, dtype=dtype
-            )
+            deepspeed.zero.Init(module=model, pin_memory=True, config=self.config, dtype=dtype)
```
[the remark this answers is not in the document]
Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:465]: I think this was just formatting done by `pre-commit`

[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/deepspeed.py:465]: Remote device was removed

[the remark this answers is not in the document]
Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:465]: woops! yep that was pesky debugging

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/deepspeed.py:636]: You should be able to tell from the stacktrace of the error

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 4]
```diff
@@ -634,9 +629,17 @@ def _auto_select_batch_size(self):
...
         if train_dl_source.is_defined():
-            train_dataloader = train_dl_source.dataloader()
-            if hasattr(train_dataloader, "batch_sampler"):
-                batch_size = train_dataloader.batch_sampler.batch_size
+            try:
+                train_dataloader = train_dl_source.dataloader()
+                if hasattr(train_dataloader, "batch_sampler"):
+                    batch_size = train_dataloader.batch_sampler.batch_size
+            except Exception:
+                if deepspeed.utils.logging.logger.level < logging.WARN:
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/deepspeed.py:637]: shouldn't this be <=?

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 5]
```diff
@@ -634,9 +629,17 @@ def _auto_select_batch_size(self):
...
+                    rank_zero_warn(
+                        "Tried to Infer the batch size for internal deepspeed logging from the `train_dataloader()`. "
+                        "To ensure DeepSpeed logging remains correct, please manually pass the plugin with the"
+                        "batch size, `Trainer(strategy=DeepSpeedPlugin(logging_batch_size_per_gpu=batch_size))`"
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/deepspeed.py:641]:
```suggestion
                        "To ensure DeepSpeed logging remains correct, please manually pass the plugin with the "
                        "batch size, `Trainer(strategy=DeepSpeedPlugin(logging_batch_size_per_gpu=batch_size))`."
```

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:636]: Sure I could wrap this in `AttributeError` as in this case I try to access an Attribute that doesn't exist, however I think its hard to know what could cause the failure (could be a result of an IO error, Value error etc).

I'd like to be able to fix the case but unfortunately I don't think I can unless someone can guide me here

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 4 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:637]: Don't think so, with `logging.WARN` the logging that uses the batch size are supressed

**Review by kaushikb11 — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 6]
```diff
@@ -312,6 +312,7 @@ def __init__(
             )
         self._config_initialized = False
         deepspeed.utils.logging.logger.setLevel(logging_level)
+        pl._logger.setLevel(logging_level)
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/training_type/deepspeed.py:315]: @carmocca not sure how I feel about this, setting the global logger level is fine? I liked setting just the deepspeed logger level as it was unintrusive to other messages

Comment (carmocca) [reply pytorch_lightning/plugins/training_type/deepspeed.py:315]: Wouldn't it make sense that if the user asks to configure the deepspeed logger, we configure ours too? Us checking the current loglevel is ugly and should rely on the logging internals.

What's missing is resetting the log level (hence failing tests)
